keps/sig-auth/20190116-service-account-external-signing.md

```diff
@@ -0,0 +1,152 @@
+
+### Updates to API server token generation
+
+As of Kubernetes v1.13.2, the API server uses the functions `JWTTokenGenerator` and `JWTTokenAuthenticator`. New types that implement the `TokenGenerator` interface and support token validation will be added to `k8s.io/kubernetes/pkg/serviceaccount/`.
+
+### New API
+
+I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+
+```proto
+syntax = "proto3";
+
+package v1beta1;
+
+service KeyService {
+  // Sign an incoming payload
+  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
+  // List supported Algorithms
+  rpc ListAlgorithms(ListAlgorithmsRequest) returns (ListAlgorithmsResponse) {}
+  // Get a specific key by ID, defaulting to the current public key
+  rpc GetPublicKey(GetPublicKeyRequest) returns (GetPublicKeyResponse) {}
+  // List all active public keys
+  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+}
+
+message SignPayloadRequest {
+  // payload is the content to be signed
+  bytes payload = 1;
+  // algorithm specifies which algorithm to sign with
+  string algorithm = 2;
+}
+message SignPayloadResponse {
+  // content returns the signed payload
+  bytes content = 1;
+}
+
+message ListAlgorithmsRequest {}
+message ListAlgorithmsResponse {
+  // signature_algorithms returns the supported algorithms
+  repeated string signature_algorithms = 1;
+}
+
+message PublicKey {
+  // public_key is a PEM encoded public key
+  bytes public_key = 1;
+  // certificate is a concatenated list of PEM encoded x509 certificates
+  bytes certificates = 2;
+  // key_id is the key's ID
+  string key_id = 3;
+  // algorithm states the algorithm the key uses
+  string algorithm = 4;
+  // use states what the key is used for
+  string use = 5 ;
+}
+
+message GetPublicKeyRequest {
+  // key_id is an optional field specifying which key to request
+  // Defaults to the current signing key
+  string key_id = 1;
+}
+message GetPublicKeyResponse {
+  // public_key contains the public key for verification
+  PublicKey public_key = 1;
+}
+
+message ListPublicKeysRequest {}
+message ListPublicKeysResponse {
+  // public_keys is a list of public verifying keys
+  repeated PublicKey public_keys = 1;
+}
+```
+
+### Implementation Details/Notes/Constraints
+
+The API server flag `--service-account-key-file` can be specified multiple times. If all files are regular files, the existing behavior will be unchanged. If both a unix socket and files are provided, the APIserver will exit with an error. If the file is a unix socket, all verifying keys for projected tokens will be read over the socket.
```

Comment:
> can be specified multiple times

This was probably a mistake...

Reply:
> This was probably a mistake in retrospect...

how so? it supports rotation methods where the old and new key are in different files?

Reply:
What value does the flexibility provide vs requiring a single file with multiple pem blocks? I would be curious to see how many clusters are taking advantage of this.